test/e2e/trafficcontrol_test.go

```diff
@@ -0,0 +1,254 @@
+}
+
+func testMirrorToRemote(t *testing.T, data *TestData) {
+	// Create a VXLAN tunnel on the collector Pod to receive mirrored packets.
+	tunnelPeer := "vxlan0"
+	cmd := fmt.Sprintf(`ip link add %[3]s type vxlan id %[1]d dstport %[2]d dev eth0 && \
+ip link set %[3]s up`, vni, dstVXLANPort, tunnelPeer)
+	_, _, err := data.RunCommandFromPod(data.testNamespace, testCollectorPodName, agnhostContainerName, []string{"sh", "-c", cmd})
+	require.NoError(t, err, "Failed to create VXLAN tunnel")
+
+	// Create a TrafficControl whose target port is VXLAN.
+	targetPort := &v1alpha2.UDPTunnel{RemoteIP: testCollectorPodIP, VNI: &vni, DestinationPort: &dstVXLANPort}
+
+	tc := data.createTrafficControl(t, "tc-", nil, labels, v1alpha2.DirectionBoth, v1alpha2.ActionMirror, targetPort, true, nil)
+	defer data.crdClient.CrdV1alpha2().TrafficControls().Delete(context.TODO(), tc.Name, metav1.DeleteOptions{})
+	// Wait flows of TrafficControl to be realized.
+	time.Sleep(time.Second)
+
+	// Verify the mirrored packets.
+	verifyMirroredPackets(t, data, tunnelPeer, true)
+}
+
+func testMirrorToLocal(t *testing.T, data *TestData) {
+	// Create a TrafficControl whose target port is OVS internal port.
+	portName := "test-port"
+	targetPort := &v1alpha2.OVSInternalPort{Name: portName}
+	tc := data.createTrafficControl(t, "tc-", nil, labels, v1alpha2.DirectionBoth, v1alpha2.ActionMirror, targetPort, false, nil)
+	defer data.crdClient.CrdV1alpha2().TrafficControls().Delete(context.TODO(), tc.Name, metav1.DeleteOptions{})
+	// Wait flows of TrafficControl to be realized.
+	time.Sleep(time.Second)
+
+	// Verify the mirrored packets.
+	verifyMirroredPackets(t, data, portName, false)
+}
+
+func testRedirectToLocal(t *testing.T, data *TestData) {
+	targetPortName := "target-veth"
+	returnPortName := "return-veth"
+	addVethPairCmd := fmt.Sprintf(`ip link add dev %[1]s type veth peer name %[2]s && \
+ip link set dev %[1]s up && \
+ip link set dev %[2]s up`, targetPortName, returnPortName)
+	delVethPairCmd := fmt.Sprintf(`ip link del dev %s`, targetPortName)
+
+	_, _, _, err := data.RunCommandOnNode(testNode, addVethPairCmd)
+	require.NoError(t, err, "Failed to create veth pair on test Node")
+	defer data.RunCommandOnNode(testNode, delVethPairCmd)
+
+	targetPort := &v1alpha2.NetworkDevice{Name: targetPortName}
+	returnPort := &v1alpha2.NetworkDevice{Name: returnPortName}
+	tc := data.createTrafficControl(t, "tc-", nil, labels, v1alpha2.DirectionBoth, v1alpha2.ActionRedirect, targetPort, false, returnPort)
+	defer data.crdClient.CrdV1alpha2().TrafficControls().Delete(context.TODO(), tc.Name, metav1.DeleteOptions{})
+	// Wait flows of TrafficControl to be realized.
+	time.Sleep(time.Second)
+
+	_, hostname, _, err := data.RunCommandOnNode(testNode, fmt.Sprintf("curl --connect-timeout 1 --retry 5 --retry-connrefused http://%s:8080/hostname", testPodIP))
+	require.NoError(t, err)
+	require.Equal(t, testPodName, hostname)
```

Comment:
Does this verification achieve the purpose? I think even TrafficControl doesn't work, the test will pass?
Should it also check the packet count on the two interfaces and the connection is still working?

Reply:
Make sense. In new test, the number of packets on test Pod, target port, return port are all checked.